vllm/engine/multiprocessing/engine.py

```diff
@@ -272,13 +291,32 @@ def _handle_abort_request(self, request: RPCAbortRequest):
         if self.log_requests:
             logger.info("Aborted request %s.", request.request_id)
 
-    def _handle_health_request(self):
+    def _heartbeat_loop(self):
+        while not self._heartbeat_stop_event.wait(
+                timeout=self.heartbeat_interval_seconds):
+            # Loops until the stop event is set
+            self._heartbeat()
+
+        logger.debug("Exiting MQLLMEngine heartbeat thread")
+
+    def _heartbeat(self):
+        # Send unhealthy if engine has already errored
         if self._errored_with is not None:
             self._send_unhealthy(self._errored_with)
 
+        # Check for life of the main loop
+        if time.time() - self._last_alive_time > self.last_alive_threshold:
+            self._send_unhealthy(RuntimeError("Engine loop has died"))
+
         # Raises error if unhealthy.
-        self.engine.check_health()
-        self._send_healthy()
+        try:
+            self.engine.check_health()
+            self._send_healthy()
+        except Exception as e:
+            # TODO: set errored_with here?
+            # Does it matter if we're just gonna kill the engine?
```

Comment:
The logic behind having `_errored_with` is such that we can put the root cause of the issue into the logs.

Reply:
Ah, yeah then it make sense to set here!